File: linked_list/21_merge_two_sorted_lists.go
type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	first := list1
	second := list2
	dummy := &ListNode{}
	current := dummy

	for first != nil && second != nil {
		if first.Val > second.Val {
			current.Next = second
			second = second.Next
		} else {
			current.Next = first
			first = first.Next
		}

		current = current.Next
	}

	// можно проще
	for first != nil {
		current.Next = first
		first = first.Next
		current = current.Next
	}
	for second != nil {
		current.Next = second
		second = second.Next
		current = current.Next
	}

	// потому что если что-то осталось то только в одном списке
	// а так как он уже отсортирован мы просто указываем на то место где остановились
	// и там гарантировано значения больше и в правильном порядке
	if first != nil {
		current.Next = first
	}

	if first != nil {
		current.Next = second
	}

	return dummy.Next
}

File: linked_list/21_merge_two_sorted_lists.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def mergeTwoLists(list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        current = dummy

        first = list1
        second = list2

        while first and second:
            if first.val < second.val:
                current.next = first
                first = first.next
            else:
                current.next = second
                second = second.next
            current = current.next
        
        while first:
            current.next = first
            first = first.next
            current = current.next
        while second:
            current.next = second
            second = second.next
            current = current.next

		# можно проще т.к. если что-то осталось то только в одном списке и поэтому гарантировано в правильном порядке
        if first:
            current.next = first
        if second:
            current.next = second

        return dummy.next